File: neetcode/10. backtracking/18.n-QueensII.py
# The n-queens puzzle is the problem of placing n queens on an n x n chessboard such 
# that no two queens attack each other.

# Given an integer n, return the number of distinct solutions to the n-queens puzzle.

from typing import List

class Solution:
  def totalNQueens(self, n: int) -> int:
    answer = [0] # use list to pass by reference
    
    cols = set() # use set to check if a column is occupied
    posdiag = set() # use set to check if a positive diagonal is occupied
    negdiag = set() # use set to check if a negative diagonal is occupied

    self.backtrack(0, n, answer, cols, posdiag, negdiag)
    return answer[0]

  def backtrack(self, i: int, n: int, answer: List[int], cols: set, posdiag: set, negdiag: set) -> None:
    if i == n:
      answer[0] += 1
      return
    
    for j in range(n):
      if j in cols or (i+j) in posdiag or (i-j) in negdiag:
        continue

      cols.add(j)
      posdiag.add(i+j)
      negdiag.add(i-j)

      self.backtrack(i+1, n, answer, cols, posdiag, negdiag)

      cols.remove(j)
      posdiag.remove(i+j)
      negdiag.remove(i-j)

File: neetcode/10. backtracking/18.n-QueensII.cpp
// The n-queens puzzle is the problem of placing n queens on an n x n chessboard such 
// that no two queens attack each other.

// Given an integer n, return the number of distinct solutions to the n-queens puzzle.

#include <set>

class Solution {
public:
  int totalNQueens(int n) {
    backtrack(0, n);
    return answer;
  }
private:
  int answer = 0;
  std::set<int> cols;
  std::set<int> posdiag;
  std::set<int> negdiag;
  void backtrack(int i, int n) {
    if (i == n) {
      answer++;
      return;
    }

    for (int j = 0; j < n; j++) {
      if (cols.find(j) != cols.end() || posdiag.find(i+j) != posdiag.end() || negdiag.find(i-j) != negdiag.end()) {
        continue;
      }

      cols.insert(j);
      posdiag.insert(i+j);
      negdiag.insert(i-j);

      backtrack(i+1, n);

      cols.erase(j);
      posdiag.erase(i+j);
      negdiag.erase(i-j);
    }
  }
};